docarray/array/mixins/io/binary.py

```diff
@@ -41,73 +43,67 @@ def load_binary(
             file_ctx = open(file, 'rb')
         else:
             raise ValueError(f'unsupported input {file!r}')
-
-        from .... import Document
-
-        with file_ctx as fp:
-            d = fp.read() if hasattr(fp, 'read') else fp
-            if get_compress_ctx(algorithm=compress) is not None:
-                d = decompress_bytes(d, algorithm=compress)
-                compress = None
-
-            if protocol == 'protobuf-array':
-                from ....proto.docarray_pb2 import DocumentArrayProto
-
-                dap = DocumentArrayProto()
-                dap.ParseFromString(d)
-
-                return cls.from_protobuf(dap)
-            elif protocol == 'pickle-array':
-                return pickle.loads(d)
-            else:
-                _len = len(random_uuid().bytes)
-                _binary_delimiter = d[:_len]  # first get delimiter
-                if _show_progress:
-                    from rich.progress import track as _track
-
-                    track = lambda x: _track(x, description='Deserializing')
-                else:
-                    track = lambda x: x
-                return cls(
-                    Document.from_bytes(od, protocol=protocol, compress=compress)
-                    for od in track(d[_len:].split(_binary_delimiter))
-                )
-
+        if return_iterator:
+            return cls._load_binary_stream(file_ctx)
+        else:
+            return cls._load_binary_all(file_ctx, protocol, compress, _show_progress)
 
     @classmethod
-    def load_binary_stream(
-        cls: Type['T'],
-        file: Union[str, BinaryIO, bytes],
+    def _load_binary_stream(
```

Comment:
I thin `_load_binary_stream` should leverage `_load_binary_all` by just giving a subset of `bytes`. I think like this we can leverage optimizations from both sides